internal/users/users_test.go

```diff
@@ -121,17 +122,174 @@ func TestUpdateLocalGroups(t *testing.T) {
 				return
 			}
 
-			got := usertests.IdemnpotentOutputFromGPasswd(t, destCmdsFile)
+			got := usertests.IdempotentGPasswdOutput(t, destCmdsFile)
 			want := testutils.LoadWithUpdateFromGolden(t, got)
 			require.Equal(t, want, got, "UpdateLocalGroups should do the expected gpasswd operation, but did not")
 		})
 	}
 }
 
+func TestCleanupSystemGroups(t *testing.T) {
+	t.Parallel()
+
+	tests := map[string]struct {
+		groupFilePath string
+
+		getUsersArgs []string
+
+		wantErr bool
+	}{
+		"No-op when there are no inactive users":        {groupFilePath: "user_in_many_groups.group"},
+		"Cleans up user from group":                     {groupFilePath: "inactive_user_in_one_group.group"},
+		"Cleans up user from multiple groups":           {groupFilePath: "inactive_user_in_many_groups.group"},
+		"Cleans up multiple users from group":           {groupFilePath: "inactive_users_in_one_group.group"},
+		"Cleans up multiple users from multiple groups": {groupFilePath: "inactive_users_in_many_groups.group"},
+
+		"Error when NSS errors out":                   {groupFilePath: "inactive_user_in_one_group.group", getUsersArgs: []string{"nss-fail"}, wantErr: true},
+		"Error on missing groups file":                {groupFilePath: "does_not_exists.group", wantErr: true},
+		"Error when groups file is malformed":         {groupFilePath: "malformed_file.group", wantErr: true},
```

Comment:
another hint that stopping earlier in that case is the best course of action: here you may have cases with groups edited in the current implementation (even if likely gpasswd would fail), but you don’t have those 2 cases, checking for gpasswd output and so on…